File: nodejs/decode-file.js
const fs = require('fs');

// Parse named arguments
function parseArgs(args) {
    const parsed = {};
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--input' && i + 1 < args.length) {
            parsed.input = args[++i];
        } else if (args[i] === '--output' && i + 1 < args.length) {
            parsed.output = args[++i];
        }
    }
    return parsed;
}

const args = parseArgs(process.argv.slice(2));

if (!args.input || !args.output) {
    console.error('Usage: node decode-file.js --input <input-file> --output <output-file>');
    console.error('Sample call: node decode-file.js --input "file_encoded.txt" --output "file.zip"');
    process.exit(1);
}

const inputFile = args.input;
const outputFile = args.output;

// Check if the input file exists
if (!fs.existsSync(inputFile)) {
    console.error(`Error: File "${inputFile}" not found.`);
    process.exit(1);
}

try {
    // Read the base64 encoded text file
    const base64Text = fs.readFileSync(inputFile, 'utf8').trim();
    
    // Decode the base64 content to binary
    const decodedBuffer = Buffer.from(base64Text, 'base64');
    
    // Save the decoded binary to the output file
    fs.writeFileSync(outputFile, decodedBuffer);
    
    console.log(`Successfully decoded "${inputFile}"`);
    console.log(`Output saved to: "${outputFile}"`);
    console.log(`Encoded size: ${base64Text.length} characters`);
    console.log(`Decoded size: ${decodedBuffer.length} bytes`);
} catch (error) {
    console.error(`Error processing file: ${error.message}`);
    process.exit(1);
}

File: python/decode-file.py
import sys
import os
import base64
import argparse


def decode_file(input_file, output_file):
    """Decode a base64-encoded text file back to its original binary form.

    Args:
        input_file: Path to the base64-encoded text file.
        output_file: Path to save the decoded binary output.

    Returns:
        A dict with 'input_file', 'output_file', 'encoded_size', and 'decoded_size'.

    Raises:
        FileNotFoundError: If the input file does not exist.
        IOError: If reading or writing fails.
    """
    if not os.path.exists(input_file):
        raise FileNotFoundError(f'File "{input_file}" not found.')

    # Read the base64 encoded text file
    with open(input_file, 'r', encoding='utf-8') as f:
        base64_text = f.read().strip()

    # Decode the base64 content to binary
    decoded_data = base64.b64decode(base64_text)

    # Save the decoded binary to the output file
    with open(output_file, 'wb') as f:
        f.write(decoded_data)

    return {
        'input_file': input_file,
        'output_file': output_file,
        'encoded_size': len(base64_text),
        'decoded_size': len(decoded_data),
    }


def main():
    parser = argparse.ArgumentParser(
        description='Decode a base64-encoded text file back to its original binary form.',
        epilog='Sample call: python decode-file.py --input photo_encoded.txt --output photo_restored.png',
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    parser.add_argument('--input', required=True, help='Path to the base64-encoded text file.')
    parser.add_argument('--output', required=True, help='Path to save the decoded output file.')
    args = parser.parse_args()

    try:
        result = decode_file(args.input, args.output)
        print(f'Successfully decoded "{result["input_file"]}"')
        print(f'Output saved to: "{result["output_file"]}"')
        print(f'Encoded size: {result["encoded_size"]} characters')
        print(f'Decoded size: {result["decoded_size"]} bytes')
    except Exception as e:
        print(f'Error: {e}', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()
